enhancements/20230830-managed-dhcp.md

```diff
@@ -0,0 +1,318 @@
+
+For users directly operate with the Kubernetes APIs using command line tools like `kubectl`, creating IPPool objects has the same effect as creating VM networks with **Managed DHCP** tab selected and configured. To allocate IP addresses for the VM's network interfaces, a corresponding VirtualMachineNetworkConfig object has to be created, filled with the network interface's MAC address and the associated network name (the name of the NetworkAttachmentDefinition,) plus the desired IP address (non-required.) Behind the scene, the controller continuously reconciles the IPPool and VirtualMachineNetworkConfig objects. When it sees an IPPool object, it spins up a DHCP server with the specified configuration and maintains it until the object is removed. When it sees a VirtualMachineNetworkConfig object, it tries to allocate IP addresses from the target IPPool object.
+
+Removing the VM will cause the associated VirtualMachineNetworkConfig object to be purged. Users can remove it directly without removing the VM. The static DHCP lease for the VM's network interfaces is removed. After the lease time, the VM loses the IP address. The same logic applies to the IPPool objects: removing the VM network (NetworkAttachmentDefinition) will cause the associated IPPool object to be purged. (Note: The IPPool object cannot be removed if there are still VirtualMachineNetworkConfig objects associated with it.) Users can remove it directly without removing the VM network. In that case, the VM network becomes a purely L2 network, just like we used to see.
+
+### API changes
+
+Two CRDs are introduced:
+
+- IPPool (ipl)
+- VirtualMachineNetworkConfig (vmnetcfg)
+
+**IPPool** is where users can define every details of the pool that serves the network.
+
+```go
+var (
+	Registered condition.Cond = "Registered"
+	CacheReady condition.Cond = "CacheReady"
+	AgentReady condition.Cond = "AgentReady"
+	Stopped    condition.Cond = "Stopped"
+)
+
+type IPPool struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec   IPPoolSpec   `json:"spec,omitempty"`
+	Status IPPoolStatus `json:"status,omitempty"`
+}
+
+type IPPoolSpec struct {
+	IPv4Config  IPv4Config `json:"ipv4Config,omitempty"`
+	NetworkName string     `json:"networkName,omitempty"`
+}
+
+type IPv4Config struct {
+	ServerIP string `json:"serverIP,omitempty"`
+	CIDR     string `json:"cidr,omitempty"`
+	Pool     Pool   `json:"pool,omitempty"`
+	// +optional
+	Router string `json:"router,omitempty"`
+	// +optional
+	DNS []string `json:"dns,omitempty"`
+	// +optional
+	DomainName *string `json:"domainName,omitempty"`
+	// +optional
+	DomainSearch []string `json:"domainSearch,omitempty"`
+	// +optional
+	NTP []string `json:"ntp,omitempty"`
+	// +optional
+	LeaseTime *int `json:"leaseTime,omitempty"`
+}
+
+type Pool struct {
+	Start string `json:"start,omitempty"`
+	End   string `json:"end,omitempty"`
+	// +optional
+	Exclude []string `json:"exclude,omitempty"`
```

Comment:
If a user deliberately crafted two vmnetcfg objects with the same designated IP address, only one of them could allocate it successfully (assuming the IP address was not in use beforehand). The other one will not get the same IP address; it will log down the error message in the condition field:

```
...
spec:
  networkConfig:
  - ipAddress: 192.168.48.82
    macAddress: 00:1a:2b:3c:4d:5e
    networkName: default/net-48
  vmName: test-vm-2
status:
  conditions:
  - lastUpdateTime: "2024-01-30T13:24:17Z"
    message: designated ip 192.168.48.82 is already allocated
    reason: Error
    status: "False"
    type: Allocated
  - lastUpdateTime: "2024-01-30T13:24:17Z"
    status: "False"
    type: Disabled
```

The IP allocation process happens on the control plane. The controller's mission is to determine what IP to allocate and ensure no double allocation. The allocation is recorded in the ippool object's status field. It could be taken as DHCP's lease database, and it will be the single source of truth of the allocation status of that specific subnet. So, it's possible to implement a validator in webhook to check whether an IP is allocated or not before creating the vmnetcfg object with a designated IP address (it hasn't been implemented yet).